### PR TITLE
Fix the newline with object literals bug

### PR DESCRIPTION
**Pull request by fatfisz**
From comment https://github.com/yannickcr/eslint-plugin-react/issues/857#issuecomment-299698993:

> Found another bug:
>
> ```
> <App foo={
> { bar: true, baz: true }
> } />;
> ```
> This would get 2 errors using the default settings: `There should be no space after '{'` and `There should be no space before '}'`.
>
> This is because conditions for the object literal spacing were written so that the second one was never reached:
>
> ```
>           if (sourceCode.isSpaceBetweenTokens(first, second)) {
>             reportNoBeginningSpace(node, first);
>           } else if (!config.allowMultiline && isMultiline(first, second)) {
>             reportNoBeginningNewline(node, first, config);
>           }
> ```
> If the first condition fails, `isMultiline` will return `false` - there is no newline.
>
> Will fix this and add appropriate tests.

### CODE REVIEW

File: tests/lib/rules/jsx-curly-spacing.js [hunk 1]
```diff
@@ -45,13 +60,21 @@ ruleTester.run('jsx-curly-spacing', rule, {
...
+    options: ['never'],
+    parserOptions: parserOptions
+  }, {
     code: '<App foo={ {bar:baz} } />;',
     options: ['always'],
     parserOptions: parserOptions
   }, {
     code: [
       '<App foo={',
-      'bar',
```
Comment (ljharb) [inline tests/lib/rules/jsx-curly-spacing.js]: can we keep this test case as well?

File: tests/lib/rules/jsx-curly-spacing.js [hunk 2]
```diff
@@ -217,6 +248,62 @@ ruleTester.run('jsx-curly-spacing', rule, {
...
+      '} />;'
+    ].join('\n'),
+    output: '<App foo={ { bar: true, baz: true } } />;',
+    options: ['always', {allowMultiline: false, spacing: {objectLiterals: 'always'}}],
+    errors: [{
+      message: 'There should be no newline after \'{\''
+    }, {
+      message: 'There should be no newline before \'}\''
+    }],
+    parserOptions: parserOptions
```
Comment (ljharb) [inline tests/lib/rules/jsx-curly-spacing.js]: Now that we support node 4+, all these can just be `parserOptions` (only the new examples; we'll change all the rest in their own commit)

File: tests/lib/rules/jsx-curly-spacing.js (hunk 1 above)
Comment (fatfisz) [inline tests/lib/rules/jsx-curly-spacing.js]: This test is identical with the one just below it, so I suggest it stays removed.

File: tests/lib/rules/jsx-curly-spacing.js (hunk 2 above)
Comment (fatfisz) [inline tests/lib/rules/jsx-curly-spacing.js]: I'm not sure if having only a part the code conform to the new convention is a good idea, especially since this convention can be applied automatically. How about I make a separate PR which replaces all occurences of `parserOptions: parserOptions` with `parserOptions` everywhere?

Comment (ljharb) [inline tests/lib/rules/jsx-curly-spacing.js]: Sure, that's fine, I suppose.

File: tests/lib/rules/jsx-curly-spacing.js (hunk 1 above)
Comment (fatfisz) [inline tests/lib/rules/jsx-curly-spacing.js]: (or is there any reason for the duplicated test?)

**Comment by fatfisz** [comment]
I've rebased this. There's only the unresolved problem of the duplicate test left.

File: tests/lib/rules/jsx-curly-spacing.js (hunk 1 above)
Comment (ljharb) [inline tests/lib/rules/jsx-curly-spacing.js]: good call, i didn't realize it was duplicated
